File: plot_3d_surface_with_raw.py
"""
Plot 3D PK-PD surface with raw data overlay for Bortezomib.
Matches the style from Paper_Plots_PKPD_Elimination_Surfaces.ipynb
"""

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use('Agg')
import matplotlib.font_manager as fm
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from pathlib import Path

# Register Helvetica fonts from local fonts folder
_font_dir = Path(__file__).parent / 'fonts'
if _font_dir.exists():
    for _font_file in _font_dir.glob('*.ttf'):
        fm.fontManager.addfont(str(_font_file))
plt.rcParams['font.family'] = 'sans-serif'
plt.rcParams['font.sans-serif'] = ['Helvetica', 'Arial', 'DejaVu Sans']
plt.rcParams['axes.unicode_minus'] = False

# =============================================================================
# CONFIGURATION
# =============================================================================

DRUG_NAME = 'Bortezomib'
PROJECT_ROOT = Path(__file__).parent
COEFF_PATH = PROJECT_ROOT / 'EQN_Coefficients' / 'all_equations_coefficients.xlsx'
O2_DATA_PATH = PROJECT_ROOT / 'Cleaned_Data' / 'O2_Mean_Averaged.xlsx'
CONTRACTILITY_DATA_PATH = PROJECT_ROOT / 'Cleaned_Data' / 'Heart_Contractility_Averaged.xlsx'
OUTPUT_DIR = PROJECT_ROOT / 'Output' / '3D_Plots'
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

# =============================================================================
# PK-PD ELIMINATION EQUATION (same as notebook)
# =============================================================================

def pkpd_elimination_response(dose_ratio, time, R0, Emax, kappa, n, m, tau, k_elim):
    """Calculate PK-PD elimination response."""
    dose_ratio = np.asarray(dose_ratio)
    time = np.asarray(time)

    # Create meshgrid if inputs are 1D
    if dose_ratio.ndim == 1 and time.ndim == 1:
        dose_ratio, time = np.meshgrid(dose_ratio, time)

    kappa = max(kappa, 1e-9)
    tau = max(tau, 1e-9)
    k_elim = max(k_elim, 1e-9)
    time = np.maximum(time, 1e-9)

    C_norm = dose_ratio * np.exp(-k_elim * time)
    time_component = (time / tau) ** m
    conc_component = C_norm ** n
    effect_term = kappa * conc_component * time_component
    response = R0 + Emax * (1 - np.exp(-effect_term))

    return response


# =============================================================================
# DATA LOADING
# =============================================================================

def load_coefficients(drug_name, response_type):
    """Load PK-PD coefficients for a drug."""
    df = pd.read_excel(COEFF_PATH, sheet_name='pkpd_elimination', header=1)
    df.columns = df.columns.str.strip()
    row = df[df['Drug'] == drug_name].iloc[0]

    suffix = '.1' if response_type == 'O2' else ''

    return {
        'R0': row[f'R0{suffix}'],
        'Emax': row[f'Emax{suffix}'],
        'kappa': row[f'kappa{suffix}'],
        'n': row[f'n{suffix}'],
        'm': row[f'm{suffix}'],
        'tau': row[f'tau{suffix}'],
        'k_elim': row[f'k_elim{suffix}'],
        'Cmax': row[f'Cmax_used{suffix}']
    }


def load_raw_data(drug_name, response_type):
    """Load raw data and convert to (time, dose_ratio, response) arrays."""
    data_path = O2_DATA_PATH if response_type == 'O2' else CONTRACTILITY_DATA_PATH
    df = pd.read_excel(data_path, sheet_name=drug_name)

    # Get Cmax for dose ratio conversion
    params = load_coefficients(drug_name, response_type)
    cmax = params['Cmax']

    # First column is time
    time_vals = df.iloc[:, 0].values

    # Remaining columns are concentrations
    conc_columns = df.columns[1:]

    # Build arrays for 3D scatter
    all_times = []
    all_dose_ratios = []
    all_responses = []

    for conc_col in conc_columns:
        try:
            conc = float(conc_col)
            dose_ratio = conc / cmax
            responses = df[conc_col].values

            for t, r in zip(time_vals, responses):
                if not np.isnan(r):
                    all_times.append(t)
                    all_dose_ratios.append(dose_ratio)
                    all_responses.append(r)
        except ValueError:
            continue

    return np.array(all_times), np.array(all_dose_ratios), np.array(all_responses)


# =============================================================================
# PLOTTING (matching notebook style)
# =============================================================================

def plot_3d_surface_with_raw(drug_name, response_type, remove_r0_offset=True):
    """
    Plot 3D surface with raw data overlay.
    Matches Paper_Plots_PKPD_Elimination_Surfaces.ipynb style.
    """
    # Load coefficients
    params = load_coefficients(drug_name, response_type)
    R0 = params['R0']
    Emax = params['Emax']
    kappa = params['kappa']
    n = params['n']
    m = params['m']
    tau = params['tau']
    k_elim = params['k_elim']

    print(f"\n{drug_name} - {response_type}")
    print(f"  Coefficients: R0={R0:.4f}, Emax={Emax:.4f}, kappa={kappa:.4f}")
    print(f"                n={n:.4f}, m={m:.4f}, tau={tau:.4f}, k_elim={k_elim:.4f}")
    print(f"  Cmax={params['Cmax']}")

    # Grid for surface (same as notebook: 60x60)
    dose_ratio = np.linspace(0, 2, 60)
    time = np.linspace(0, 96, 60)
    T, Dr = np.meshgrid(time, dose_ratio)

    # Calculate response surface
    Response = pkpd_elimination_response(Dr, T, R0, Emax, kappa, n, m, tau, k_elim)
    if remove_r0_offset:
        Response = Response - R0

    # Load raw data
    raw_times, raw_dose_ratios, raw_responses = load_raw_data(drug_name, response_type)
    if remove_r0_offset:
        raw_responses = raw_responses - R0

    print(f"  Raw data: {len(raw_times)} points")
    print(f"  Dose ratio range in raw data: {raw_dose_ratios.min():.2f} - {raw_dose_ratios.max():.2f}")

    # Create figure
    fig = plt.figure(figsize=(10, 8))
    ax = fig.add_subplot(111, projection='3d')

    # Color normalization
    vmin = 0
    vmax = np.nanmax(Response)
    norm = plt.Normalize(vmin=vmin, vmax=vmax)

    # Plot surface (same style as notebook)
    surf = ax.plot_surface(T, Dr, Response, cmap='turbo', norm=norm,
                          linewidth=0, antialiased=True, edgecolor='none', alpha=0.7)

    # Plot raw data as scatter points
    # Only plot points within the surface range (dose_ratio 0-2)
    mask = raw_dose_ratios <= 2.0
    ax.scatter(raw_times[mask], raw_dose_ratios[mask], raw_responses[mask],
               c='black', s=15, alpha=0.8, label='Raw Data', depthshade=True)

    # Also plot points outside range with different color (optional)
    mask_outside = raw_dose_ratios > 2.0
    if mask_outside.any():
        ax.scatter(raw_times[mask_outside], raw_dose_ratios[mask_outside], raw_responses[mask_outside],
                   c='red', s=15, alpha=0.6, marker='^', label='Raw Data (>2x Cmax)')

    # Set Z-axis limits
    z_min = min(0, np.nanmin(raw_responses))
    z_max = max(vmax, np.nanmax(raw_responses)) * 1.1
    ax.set_zlim(z_min, z_max)

    # Labels (same as notebook)
    ax.set_xlabel('Time (hours)', fontsize=10)
    ax.set_ylabel('Dose Ratio (C0/Cmax)', fontsize=10)
    z_label = f'{response_type} Response'
    if remove_r0_offset:
        z_label += ' (R0 removed)'
    ax.set_zlabel(z_label, fontsize=10)

    ax.set_title(f'{drug_name} - {response_type}', fontsize=12, pad=20)

    # Set view angle (same as notebook)
    ax.view_init(elev=25, azim=-158)

    # Add legend
    ax.legend(loc='upper left')

    plt.tight_layout()
    return fig


# =============================================================================
# MAIN
# =============================================================================

def main():
    print("=" * 70)
    print("3D PK-PD SURFACE WITH RAW DATA OVERLAY")
    print("=" * 70)

    for response_type in ['O2', 'Contractility']:
        fig = plot_3d_surface_with_raw(DRUG_NAME, response_type, remove_r0_offset=True)

        # Save
        filename = f"{DRUG_NAME}_{response_type}_3D_Surface_with_Raw"
        fig.savefig(OUTPUT_DIR / f"{filename}.png", dpi=300, bbox_inches='tight')
        fig.savefig(OUTPUT_DIR / f"{filename}.pdf", bbox_inches='tight')
        print(f"  Saved: {filename}.png/pdf")

        plt.close(fig)

    print("\n" + "=" * 70)
    print(f"Output directory: {OUTPUT_DIR}")
    print("=" * 70)


if __name__ == '__main__':
    main()

File: plot_3d_surface_with_raw.m
%% 3D PK-PD Surface with Raw Data Overlay
% Matches the style from Paper_Plots_PKPD_Elimination_Surfaces.ipynb
% Interactive MATLAB viewer for Bortezomib (or other drugs)

clear; clc; close all;

%% CONFIGURATION
DRUG_NAME = 'Bortezomib';
RESPONSE_TYPE = 'O2';  % 'O2' or 'Contractility'
REMOVE_R0_OFFSET = true;

% Paths (adjust if needed)
PROJECT_ROOT = fileparts(mfilename('fullpath'));
COEFF_PATH = fullfile(PROJECT_ROOT, 'EQN_Coefficients', 'all_equations_coefficients.xlsx');
O2_DATA_PATH = fullfile(PROJECT_ROOT, 'Cleaned_Data', 'O2_Mean_Averaged.xlsx');
CONTRACTILITY_DATA_PATH = fullfile(PROJECT_ROOT, 'Cleaned_Data', 'Heart_Contractility_Averaged.xlsx');

%% Load Coefficients
fprintf('Loading coefficients for %s - %s...\n', DRUG_NAME, RESPONSE_TYPE);

% Read the pkpd_elimination sheet (header on row 2)
coeff_table = readtable(COEFF_PATH, 'Sheet', 'pkpd_elimination', 'HeaderLines', 1);

% Find the drug row
drug_idx = find(strcmp(coeff_table.Drug, DRUG_NAME));
if isempty(drug_idx)
    error('Drug "%s" not found in coefficients file', DRUG_NAME);
end

% Extract coefficients based on response type
if strcmp(RESPONSE_TYPE, 'O2')
    % O2 columns have '.1' suffix (becomes '_1' in MATLAB)
    R0 = coeff_table.R0_1(drug_idx);
    Emax = coeff_table.Emax_1(drug_idx);
    kappa = coeff_table.kappa_1(drug_idx);
    n = coeff_table.n_1(drug_idx);
    m = coeff_table.m_1(drug_idx);
    tau = coeff_table.tau_1(drug_idx);
    k_elim = coeff_table.k_elim_1(drug_idx);
    Cmax = coeff_table.Cmax_used_1(drug_idx);
else
    % Contractility columns (no suffix)
    R0 = coeff_table.R0(drug_idx);
    Emax = coeff_table.Emax(drug_idx);
    kappa = coeff_table.kappa(drug_idx);
    n = coeff_table.n(drug_idx);
    m = coeff_table.m(drug_idx);
    tau = coeff_table.tau(drug_idx);
    k_elim = coeff_table.k_elim(drug_idx);
    Cmax = coeff_table.Cmax_used(drug_idx);
end

fprintf('  R0=%.4f, Emax=%.4f, kappa=%.4f\n', R0, Emax, kappa);
fprintf('  n=%.4f, m=%.4f, tau=%.4f, k_elim=%.4f\n', n, m, tau, k_elim);
fprintf('  Cmax=%.4f\n', Cmax);

%% PK-PD Elimination Equation
% R(C0, t) = R0 + Emax * (1 - exp(-kappa * (C0/Cmax * exp(-k_elim * t))^n * (t/tau)^m))

pkpd_response = @(dose_ratio, time) R0 + Emax * (1 - exp(-kappa * ...
    (dose_ratio .* exp(-k_elim * time)).^n .* (time / tau).^m));

%% Load Raw Data FIRST (to get dose ratio range)
fprintf('Loading raw data...\n');

if strcmp(RESPONSE_TYPE, 'O2')
    data_path = O2_DATA_PATH;
else
    data_path = CONTRACTILITY_DATA_PATH;
end

raw_table = readtable(data_path, 'Sheet', DRUG_NAME);

% First column is time
time_vals = raw_table{:, 1};

% Get concentration columns (numeric headers)
var_names = raw_table.Properties.VariableNames(2:end);

% Build arrays for 3D scatter
raw_times = [];
raw_dose_ratios = [];
raw_responses = [];

for i = 1:length(var_names)
    col_name = var_names{i};

    % Try to parse concentration from column name
    % MATLAB converts numeric headers like "0.05" to "x0_05"
    col_clean = strrep(col_name, 'x', '');
    col_clean = strrep(col_clean, '_', '.');
    conc = str2double(col_clean);

    if ~isnan(conc)
        dose_ratio = conc / Cmax;
        col_data = raw_table{:, i+1};  % +1 because first col is time

        for j = 1:length(time_vals)
            if ~isnan(col_data(j))
                raw_times(end+1) = time_vals(j);
                raw_dose_ratios(end+1) = dose_ratio;
                raw_responses(end+1) = col_data(j);
            end
        end
    end
end

fprintf('  Raw data: %d points\n', length(raw_times));
fprintf('  Dose ratio range: %.2f - %.2f\n', min(raw_dose_ratios), max(raw_dose_ratios));

%% Create Surface Grid (extend to cover all raw data)
max_dose_ratio = max(2, ceil(max(raw_dose_ratios)));  % At least 2, or cover all data
dose_ratio_vec = linspace(0, max_dose_ratio, 60);
time_vec = linspace(0, 96, 60);
[T, Dr] = meshgrid(time_vec, dose_ratio_vec);

% Compute response surface
Response = pkpd_response(Dr, T);

% Remove R0 offset if requested
if REMOVE_R0_OFFSET
    Response = Response - R0;
end

% Remove R0 offset from raw data if requested
if REMOVE_R0_OFFSET
    raw_responses = raw_responses - R0;
end

%% Create 3D Plot
figure('Position', [100, 100, 900, 700], 'Color', 'w');

% Plot surface
surf(T, Dr, Response, 'EdgeColor', 'none', 'FaceAlpha', 0.7);
colormap(turbo);
hold on;

% Color limits
vmin = 0;
vmax = max(Response(:));
caxis([vmin, vmax]);

% Plot all raw data points
scatter3(raw_times, raw_dose_ratios, raw_responses, ...
    25, 'k', 'filled', 'MarkerFaceAlpha', 0.8);

% Labels
xlabel('Time (hours)', 'FontSize', 11);
ylabel('Dose Ratio (C_0/C_{max})', 'FontSize', 11);
if REMOVE_R0_OFFSET
    zlabel([RESPONSE_TYPE ' Response (R_0 removed)'], 'FontSize', 11);
else
    zlabel([RESPONSE_TYPE ' Response'], 'FontSize', 11);
end

title(sprintf('%s - %s', DRUG_NAME, RESPONSE_TYPE), 'FontSize', 13, 'FontWeight', 'bold');

% Set view angle (same as notebook: elev=25, azim=-158)
view(-158, 25);

% Z-axis limits
z_min = min(0, min(raw_responses));
z_max = max(vmax, max(raw_responses)) * 1.1;
zlim([z_min, z_max]);

% Legend
legend({'Surface', 'Raw Data'}, 'Location', 'northeast');

% Colorbar
c = colorbar;
c.Label.String = 'Response';
c.Label.FontSize = 10;

% Grid
grid on;
box on;

% Enable rotation
rotate3d on;

fprintf('\nPlot created. Use mouse to rotate.\n');
fprintf('Press any key to switch response type, or close figure to exit.\n');

%% Interactive: Press key to switch between O2 and Contractility
% (Optional - comment out if not needed)
% waitforbuttonpress;
